### ISSUE TITLE
CNN on Deepsig (RF Dataset) : strange behaviour

### ISSUE THREAD

**Issue by EdgarLemaire** [CEA-LIST/N2D2#73]
Hello,

I am trying to run a CNN on the Deepsig dataset (RF dataset : https://www.deepsig.ai/datasets), and I have a very strange behaviour which I can't explain. The testing accuracy (sensitivity) is unsatisfying.

The data are 1D vectors of 1024 elements and 2 channels, which I have extracted in text files with two columns and 1024 rows. See the ini file version and an extract of such a .txt file below. The data are organized in 24 subfolders corresponding to the 24 classes of the dataset.
With the exact same topology and input data, tensorflow acheives 65% accuracy, but on N2D2, I cannot go over 20%. Moreover, after a few epochs, the testing accuracy reaches 20%, and then drops to 5% (random prediction...) and doesn't go up again. I wonder why, I have tried several hyper-parameters configuration but cannot solve the problem. I guess the problem arises from the data format, but I don't know how to do any different...

Do you have any idea on how to improve my classification accuracy, regarding dataset format or parameters ?

Thanks in Advance...

Best regards,

Edgar Lemaire

Here is an extract of a data file as a .txt file :

>  [...]
> -0.19127866625785828    0.4161767363548279
> -0.37147578597068787    0.029684072360396385
> -0.7426016330718994    -0.25420624017715454
> -0.418613076210022    -1.259751558303833
> 0.5229002833366394    -0.7010774612426758
> -0.4000523090362549    0.05423561483621597
> 0.47874999046325684    -0.8334897756576538
> 0.46665695309638977    -0.24203015863895416
> [...]

Here is my ini file, with the reshape transform that transform the 2D data into 1D 2 channels data :

> 
> DefaultModel=Frame_CUDA
> 
> 
> ; Database
> [database]
> Type=DIR_Database
> DataPath=/home/elemaire/SSD512/Deepsig
> Learn=0.8
> Validation=0.05 ; Use 20% of the dataset for validation
> 
> ; Environment
> [sp]
> SizeX=1
> SizeY=1024
> NbChannels=2
> BatchSize=64
> 
> [sp.Transformation_1]
> Type=ReshapeTransformation
> NbRows=1024
> NbCols = 1
> NbChannels = 2
> 
> 
> [conv1]
> Input=sp
> Type=Conv
> KernelDims=1 5
> NbOutputs=32
> ActivationFunction=Rectifier
> WeightsFiller=HeFiller
> ConfigSection=common.config
> 
> [pool1]
> Input=conv1
> Type=Pool
> PoolDims=1 2
> NbOutputs=[conv1]NbOutputs
> Stride=2
> Pooling=Max
> Mapping.Size=1
> 
> 
> [conv2]
> Input=pool1
> Type=Conv
> KernelDims=1 5
> NbOutputs=64
> ActivationFunction=Rectifier
> WeightsFiller=HeFiller
> ConfigSection=common.config
> 
> [pool2]
> Input=conv2
> Type=Pool
> PoolDims=1 2
> NbOutputs=[conv2]NbOutputs
> Stride=2
> Pooling=Max
> Mapping.Size=1
> 
> 
> 
> [conv3]
> Input=pool2
> Type=Conv
> KernelDims=1 5
> NbOutputs=128
> ActivationFunction=Rectifier
> WeightsFiller=HeFiller
> ConfigSection=common.config
> 
> [pool3]
> Input=conv3
> Type=Pool
> PoolDims=1 2
> NbOutputs=[conv3]NbOutputs
> Stride=2
> Pooling=Max
> Mapping.Size=1
> 
> 
> 
> 
> [fc1]
> Input=pool3
> Type=Fc
> NbOutputs=256
> ActivationFunction=Rectifier
> WeightsFiller=HeFiller
> ConfigSection=common.config
> 
> 
> [fc2]
> Input=fc1
> Type=Fc
> NbOutputs=128
> ActivationFunction=Rectifier
> WeightsFiller=HeFiller
> ConfigSection=common.config
> 
> 
> [fc3]
> Input=fc2
> Type=Fc
> NbOutputs=24
> ActivationFunction=Linear
> ConfigSection=common.config
> 
> [softmax]
> Input=fc3
> Type=Softmax
> NbOutputs=[fc3]NbOutputs
> WithLoss=1
> 
> [softmax.Target]
> 
> [common.config]
> NoBias=1
> WeightsSolver.LearningRate=0.05
> WeightsSolver.Momentum=0.9
> WeightsSolver.Decay=0.0005
> Solvers.LearningRatePolicy=StepDecay
> Solvers.LearningRateStepSize=[sp]_EpochSize
> Solvers.LearningRateDecay=0.993
> Solvers.Clamping=-1.0:1.0

**Comment by olivierbichler-cea** [comment]
Did you try removing "Solvers.Clamping=-1.0:1.0"?
Also, how the weights are initialized in TF?

**Comment by EdgarLemaire** [comment]
Yes, I have tried that with no success... however, I need to transcode the model toward spiking domain afterwards, thus I need to use such clamping if i'm right. In TF, the weights are initialized using "glorot uniform". I will try to change the initialization to uniform in my ini file and come back to you :)

**Comment by olivierbichler-cea** [comment]
Glorot Uniform is indeed equivaent to XavierFiller. But for ReLU activation, the HeFiller should work even better in general.

Did you check the input data of the network? Is it read correctly in N2D2?
You can check a few example in the "frames" directory, created when you run a simulation.

**Comment by EdgarLemaire** [comment]
Hello Olivier,

I have much difficulty interpreting the frame example in "frames" directory, as the data are RF signals...
In frames, for exampke frame_0, I have 2 images frame_0[0] and frame_0[1], both 2D images. frame_0[0] seems to be random, and frame_0[1] seems to be textured with diagonal lines. What does those 2 images correspond to ? Is this the input before and after transformation ? Why is it 2D and not 1D ? I guess there is indeed a problem here...

As a reminder, those are 2-channels 1D vectors of 1024 elements. Do you think a storage format in text files with 2 columns and 1024 rows, associated with the following Reshape transformation could do the trick ?

```
[sp.Transformation_1]
Type=ReshapeTransformation
NbRows=1024
NbCols = 1
NbChannels = 2
```

**Comment by EdgarLemaire** [comment]
Here is a link to a Google Drive folder which contains the "Frames" directory in case you want to take a look at it.

https://drive.google.com/drive/folders/19xXNV0JsaHnSU2rVivMh1RAb_7EBTKMf?usp=sharing

**Comment by olivierbichler-cea** [comment]
The images are always 2D, even for 1D data, there is "line-wrapping", so this is normal. However, the fact that the 2 channels are very different seems suspicious. At least the second channel is almost certainly wrong. The data is not correctly loaded.
Do you have an example of data file that you load in N2D2?

**Comment by EdgarLemaire** [comment]
OK, thank you for the explanation.
Here is a link to a data file : 
https://drive.google.com/file/d/1JE-p_6Qp1Btpl0tzqIiBnmOyRo64uFG_/view?usp=sharing

**Comment by olivierbichler-cea** [comment]
Thanks. There was an issue with the generated images in the "frames" folder that has been fixed in the latest commit.
You should now get the correct data in "frames".
However, the issue was only with the saved images, the network received the correct data...
Did you check that the weights initialization is the same as in TensorFlow? Also, are you sure there is no extra data pre-processing involved?

**Comment by EdgarLemaire** [comment]
Ok :)
Yes, the weight initialization is the same as in tensorflow as XavierFiller is equivalent to Glorot Uniform in TF.
I am sure there is no extra data pre-processing as I have saved the data in text files by truncating my TF script just before sending the data to the network and saving it instead.
The really bizarre thing is that it seems to learn at the beginning, as the recognition rate (sensitivity) climbs to 20% during the first few epochs, then rapidly drops to 4% (random level) and stays there for the rest of the training.

**Comment by EdgarLemaire** [comment]
If you want to investigate further, I can send you my python code for TF training and txt files generation, so that you can reproduce the issue on your side. Just tell me and I'll send you the files :)


Edgar

**Comment by olivierbichler-cea** [comment]
You can send me the TF code by mail and I will have a look.
Olivier
